src/hammer-vlsi/par/openroad/__init__.py

```diff
@@ -84,148 +181,959 @@ def tool_config_prefix(self) -> str:
+        report_clock_skew -digits 3
+        report_power -corner "hold"
+
+        report_floating_nets -verbose
+        report_design_area
+
+        utl::metric "worst_slack_min" [sta::worst_slack -min]
+        utl::metric "worst_slack_max" [sta::worst_slack -max]
+        utl::metric "tns_max" [sta::total_negative_slack -max]
+        utl::metric "clock_skew" [sta::worst_clock_skew -setup]
+        utl::metric "max_slew_violations" [sta::max_slew_violation_count]
+        utl::metric "max_fanout_violations" [sta::max_fanout_violation_count]
+        utl::metric "max_capacitance_violations" [sta::max_capacitance_violation_count]
+        # report clock period as a metric for updating limits
+        utl::metric "clock_period" [get_property [lindex [all_clocks] 0] period]
+        """)
+        return True
+
     def write_design(self) -> bool:
-        # TODO: currently using OpenROAD's default finish script
-        self.cmds += [
-          "make DESIGN_CONFIG={conf} -f {make} finish".format(
-            conf=self.design_config_path(), 
-            make=self.openroad_flow_makefile_path()
-        )]
+        self.block_append(f"""
+        ################################################################
+        # Write Design
+        """)
+        self.append(f"write_def {self.output_def_filename}")
+        
+        # TODO: look at IR drop analysis from ~OpenROAD-flow-scripts/flow/scripts/final_report.tcl
+        # Static IR drop analysis
+
+        # Write netlist
+        self.write_netlist()
+
+        # GDS streamout.
+        self.write_gds()
+
+        # Write SDF
+        self.write_sdf()
+
+        # Make sure that generated-scripts exists.
+        os.makedirs(self.generated_scripts_dir, exist_ok=True)
+
+        self.write_reports()
+
+        self.ran_write_design=True
+
         return True
 
+    
+
+    @staticmethod
+    def generate_chip_size_constraint(width: Decimal, height: Decimal, left: Decimal, bottom: Decimal, right: Decimal,
+                                      top: Decimal, site: str) -> str:
+        """
+        Given chip width/height and margins, generate an Innovus TCL command to create the floorplan.
+        Also requires a technology specific name for the core site
+        """
+
+        return f"initialize_floorplan -site {site} -die_area {{ 0 0 {width} {height}}} -core_area {{{left} {bottom} {width-right} {height-top}}}"
+
+    def generate_make_tracks(self) -> List[str]:
+        output = []
+        # initialize_floorplan removes existing tracks 
+        #   --> use the make_tracks command to add routing tracks 
+        #       to a floorplan (with no arguments it uses default from tech LEF)
+        layers = self.get_setting("par.generate_power_straps_options.by_tracks.strap_layers")
+        for metal in self.get_stackup().metals:
+            output.append(f"make_tracks {metal.name}")
+        return output
+
+    def create_floorplan_tcl(self) -> List[str]:
```

Comment:
It looks to me like there's a lot of stuff copied over from Innovus left in here?